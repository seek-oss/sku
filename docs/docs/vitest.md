# Vitest

[Vitest] is a testing framework that supports ESM out-of-the-box, integrates with the Vite ecosystem and has a similar API to Jest.
These features make it a great replacement for Jest in `sku` applications, especially given [Jest's current limitations with ESM].
Sku supports Vitest as an alternate to Jest since v15

**We recommend using vitest with the vite bundler and in an ESM project**, however these are not requirements.

## Migrating to Vitest

Migrating to Vitest should be fairly straightforward.
Before making any changes, it is highly recommend to read [the Vitest documentation](https://vitest.dev/guide/) to familiarise yourself with the API and features, as well as to better understand the differences between Jest and Vitest.

To enable Vitest in `sku`, first install the required dependencies:

```bash
pnpm add -D vitest @sku-lib/vitest
```

Then, configure [`testRunner`][test runner] in your `sku` config:

```typescript
// sku.config.ts
import type { SkuConfig } from 'sku';

export default {
  testRunner: 'vitest',
  ...
} satisfies SkuConfig;
```

`sku` will now invoke [the `vitest` CLI][Vitest CLI] instead of the `jest` CLI when running `sku test`.

#### Key differences between Vitest and Jest

Vitest has strong compatibility with the Jest API, however it still has differences that may affect your tests.
Differences that affect sku projects are listed here.
The full list is documented in the Vitest documentation: [Migrating from Jest to Vitest].

To automate most of the migration process, a codemod is available.
Note that additional changes may still be required after running this codemod.

```sh
pnpm dlx @sku-lib/codemod jest-to-vitest .
```

**Default Watch Mode**

`vitest` defaults to watch mode when running tests.
To run tests without watch mode you can use the `--run` flag:

```json
// package.json
{
  "scripts": {
    // ...
    "test": "sku test --run"
  }
}
```

Watch mode won't trigger in CI environments, so it's safe to omit the flag in your pipeline.

**Code Coverage Dependencies**

Vitest does not install code coverage dependencies by default.
To collect code coverage, pass the `--coverage` flag to the `sku test` command and install the `@vitest/coverage-v8` package:

```bash
pnpm add -D @vitest/coverage-v8
```

**Testing Library Matchers**

If your test setup file includes an import for `@testing-library/jest-dom`, you may need to change this to `@testing-library/jest-dom/vitest`:

```diff
// test-setup.ts
- import '@testing-library/jest-dom';
+ import '@testing-library/jest-dom/vitest';
```

**Globals Disabled**

Jest enables global APIs such as `it`, `describe`, `beforeAll`, etc., by default.
Vitest does not.
Sku adopts the Vitest defaults, meaning you will need to explicitly import these test functions.
This change results in a cleaner global namespace, better type safety, clearer dependencies and is more consistency with modern javascript practices.

```diff
+ import { describe, expect, it } from 'vitest';
```

Be aware that since globals are disabled, some common libraries like `testing-library` will not run auto DOM cleanup.
If using these libraries, you will need to add cleanup to your configured `setupTests` file.

```diff
import '@testing-library/jest-dom/vitest';

+ import { cleanup } from '@testing-library/react';
+ import { afterEach } from 'vitest';

+ afterEach(cleanup)
```

[Vitest]: https://vitest.dev/
[Jest's current limitations with ESM]: https://jestjs.io/docs/ecmascript-modules
[test runner]: ./docs/configuration.md#testrunner
[Vitest CLI]: https://vitest.dev/guide/cli.html
[codemod]: https://codemod.com/registry/jest-vitest
[Migrating from Jest to Vitest]: https://vitest.dev/guide/migration.html#jest
[Vitest globals]: https://vitest.dev/config/#globals
