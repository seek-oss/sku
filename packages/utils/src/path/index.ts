export { toPosixPath } from './toPosixPath.js';
