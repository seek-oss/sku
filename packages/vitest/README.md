[![npm](https://img.shields.io/npm/v/@sku-lib/vitest.svg?style=flat-square)](https://www.npmjs.com/package/@sku-lib/vitest)

# vitest runner for `sku test`
