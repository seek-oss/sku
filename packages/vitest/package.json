{
  "name": "@sku-lib/vitest",
  "version": "0.2.2",
  "description": "vitest runner for sku test",
  "types": "./dist/index.d.ts",
  "main": "./dist/index.js",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    },
    "./package.json": "./package.json"
  },
  "engines": {
    "node": ">=22.19.0"
  },
  "files": [
    "dist"
  ],
  "type": "module",
  "scripts": {
    "build": "tsc --project tsconfig.build.json",
    "lint:tsc": "tsc --noEmit"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/seek-oss/sku.git",
    "directory": "packages/vite"
  },
  "author": "SEEK",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/seek-oss/sku/issues"
  },
  "homepage": "https://github.com/seek-oss/sku#readme",
  "devDependencies": {
    "typescript-transform-paths": "^3.5.2",
    "vitest": "catalog:"
  },
  "dependencies": {
    "@vanilla-extract/vite-plugin": "catalog:",
    "vite-tsconfig-paths": "^5.1.4"
  },
  "peerDependencies": {
    "vitest": "^3.0.0"
  }
}
