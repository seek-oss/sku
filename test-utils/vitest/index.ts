export { getAppSnapshot } from './appSnapshot.ts';
