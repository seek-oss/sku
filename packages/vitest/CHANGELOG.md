# @sku-lib/vitest

## 0.2.1

### Patch Changes

- Fix `watch` mode ([#1393](https://github.com/seek-oss/sku/pull/1393))

## 0.2.0

### Minor Changes

- Disable Vitest globals ([#1387](https://github.com/seek-oss/sku/pull/1387))

  `vitest` is now a peer dependency. As such, `vitest` APIs must now be imported explicitly

  **EXAMPLE USAGE**:

  ```ts
  // example.test.ts
  import { describe, it, expect } from 'vitest';

  describe("test suite", () => {
    it("should be true" () => {
      expect(true).toBe(true);
    });
  });
  ```

### Patch Changes

- `deps`: Update `vite-plugin-cjs-interop` from `^2.2.0` to `^2.3.0` ([#1386](https://github.com/seek-oss/sku/pull/1386))

- Added missing `compilePackages` and `cjsInteropDependencies` support ([#1381](https://github.com/seek-oss/sku/pull/1381))

- Fixed a bug causing `sku test` to hang when `vitest` is executed in `run` mode ([#1387](https://github.com/seek-oss/sku/pull/1387))

## 0.1.1

### Patch Changes

- Experimental vite bundler now requires vite v7. ([#1346](https://github.com/seek-oss/sku/pull/1346))

  This is a breaking peer dependency change for projects that use the experimental vite bundler. Since this is an experimental feature, we are not bumping the major version.

- - `vitest`: Honour watch mode when running `sku test` ([#1325](https://github.com/seek-oss/sku/pull/1325))

    Previously it was not possible to run Vitest in watch mode via `sku test`. `sku test` should now behave similarly to [the `vitest` CLI][vitest cli], so `sku test`, `sku test watch` and `sku test -w/--watch` will all trigger watch mode.

    To run tests once and exit, you can execute `sku test run`, `sku test --run` or `sku test --no-watch` .

    [vitest cli]: https://vitest.dev/guide/cli

  - `deps` Bump peer dependencies to support React 19

## 0.1.0

### Minor Changes

- Initial release. ([#1273](https://github.com/seek-oss/sku/pull/1273))
