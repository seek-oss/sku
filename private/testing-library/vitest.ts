// Re-export vitest setup from cli-testing-library
import 'cli-testing-library/vitest';
