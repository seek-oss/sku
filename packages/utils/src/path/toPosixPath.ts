import path from 'node:path';

/**
 * Replaces all win32 path separators (\) with posix path separators (/)
 */
export const toPosixPath = (inputPath: string) =>
  inputPath.replaceAll(path.win32.sep, path.posix.sep);
