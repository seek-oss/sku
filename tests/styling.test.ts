import { describe, beforeAll, afterAll, it, expect } from 'vitest';
import { getAppSnapshot } from '@sku-private/vitest-utils';
import assert from 'node:assert/strict';
import path from 'node:path';
import type { Page } from 'puppeteer';
import {
  dirContentsToObject,
  waitForUrls,
  runSkuScriptInDir,
  getStoryPage,
  getTextContentFromFrameOrPage,
  gracefulSpawn,
} from '@sku-private/test-utils';
import type { ChildProcess } from 'node:child_process';
import skuConfigImport from '../fixtures/styling/sku.config.ts';

import { createRequire } from 'node:module';

const require = createRequire(import.meta.url);

const appDir = path.dirname(
  require.resolve('@sku-fixtures/styling/sku.config.ts'),
);
const distDir = path.resolve(appDir, 'dist');

const skuConfig = skuConfigImport as unknown as typeof skuConfigImport.default;

assert(skuConfig.port, 'sku config has port');
const devServerUrl = `http://localhost:${skuConfig.port}`;

describe('styling', () => {
  describe('build', () => {
    let process: ChildProcess;

    beforeAll(async () => {
      await runSkuScriptInDir('build', appDir);
      process = await runSkuScriptInDir('serve', appDir);
      await waitForUrls(devServerUrl);
    });

    afterAll(async () => {
      await process.kill();
    });

    it('should create valid app', async () => {
      const app = await getAppSnapshot(devServerUrl);
      expect(app).toMatchSnapshot();
    });

    it('should generate the expected files', async () => {
      const files = await dirContentsToObject(distDir);
      expect(files).toMatchSnapshot();
    });
  });

  describe('start', () => {
    let server: ChildProcess;

    beforeAll(async () => {
      server = await runSkuScriptInDir('start', appDir);
      await waitForUrls(devServerUrl);
    });

    afterAll(async () => {
      await server.kill();
    });

    it('should start a development server', async () => {
      const snapshot = await getAppSnapshot(devServerUrl);
      expect(snapshot).toMatchSnapshot();
    });
  });

  describe('test', () => {
    let exitCode: number | null;

    beforeAll(async () => {
      const { child } = await runSkuScriptInDir('test', appDir);
      exitCode = child.exitCode;
    });

    it('should handle Vanilla Extract styles in tests', async () => {
      expect(exitCode).toEqual(0);
    });
  });

  describe('storybook', () => {
    const storybookPort = 8090;
    const storybookBaseUrl = `http://localhost:${storybookPort}`;

    const storyIframePath = '/iframe.html?viewMode=story&id=blueblock--default';
    const storyIframeUrl = `${storybookBaseUrl}${storyIframePath}`;

    let server: ChildProcess;
    let storyPage: Page;

    beforeAll(async () => {
      server = gracefulSpawn(
        'pnpm',
        [
          'storybook',
          'dev',
          '--ci',
          '--quiet',
          '--port',
          storybookPort.toString(),
        ],
        {
          cwd: appDir,
          stdio: 'inherit',
        },
      );
      await waitForUrls(storyIframeUrl);
      storyPage = await getStoryPage(storyIframeUrl);
    }, 200000);

    afterAll(async () => {
      await server.kill();
    });

    it('should render external styles', async () => {
      const { text, fontSize } = await getTextContentFromFrameOrPage(
        storyPage,
        '[data-automation-external]',
      );

      expect(text).toEqual('This should be invisible');
      expect(fontSize).toEqual('9px');
    });

    it('should render Vanilla Extract styles', async () => {
      const { fontSize } = await getTextContentFromFrameOrPage(
        storyPage,
        '[data-automation-vanilla]',
      );

      expect(fontSize).toEqual('64px');
    });
  });
});
