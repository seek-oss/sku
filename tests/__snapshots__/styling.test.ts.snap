// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styling build should create valid app 1`] = `
SCRIPTS: [
  "/styling/runtime.js",
  "/styling/2.js",
  "/styling/main.js",
]
CSS: [
  "/styling/main-464f06daf5715e3d7e00.css",
]
SOURCE HTML: <!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      My Awesome Project
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      data-chunk="main"
      rel="stylesheet"
      href="styles[0]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[0]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[1]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[2]"
    >
  </head>
  <body>
    <div id="app">
      <div>
        <div>
          Render type:
          Initial
        </div>
        <div class="ss2a6DC">
          <span
            class="external"
            data-automation-external="true"
          >
            This should be invisible
          </span>
          <div
            class="x5z4AMj"
            data-automation-less="true"
          >
          </div>
          <div
            class="qiwkl41 qiwkl40"
            data-automation-vanilla="true"
          >
          </div>
        </div>
      </div>
    </div>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS__"
      type="application/json"
    >
      []
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS___ext"
      type="application/json"
    >
      {"namedChunks":[]}
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[0]"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[1]"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[2]"
    >
    </script>
  </body>
</html>
POST HYDRATE DIFFS: 
       <div>
         <div>
           Render type:
-          Initial
+          Client
         </div>
         <div class="ss2a6DC">
           <span

`;

exports[`styling build should generate the expected files 1`] = `
{
  "2.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "components/styles.less.d.ts": "// This file is automatically generated.
// Please do not change this file!
interface CssExports {
  'nested': string;
  'root': string;
}
export const cssExports: CssExports;
export default cssExports;
",
  "index.html": SCRIPTS: [
    "/styling/runtime.js",
    "/styling/2.js",
    "/styling/main.js",
  ]
CSS: [
    "/styling/main-464f06daf5715e3d7e00.css",
  ]
SOURCE HTML: <!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      My Awesome Project
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      data-chunk="main"
      rel="stylesheet"
      href="styles[0]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[0]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[1]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[2]"
    >
  </head>
  <body>
    <div id="app">
      <div>
        <div>
          Render type:
          Initial
        </div>
        <div class="ss2a6DC">
          <span
            class="external"
            data-automation-external="true"
          >
            This should be invisible
          </span>
          <div
            class="x5z4AMj"
            data-automation-less="true"
          >
          </div>
          <div
            class="qiwkl41 qiwkl40"
            data-automation-vanilla="true"
          >
          </div>
        </div>
      </div>
    </div>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS__"
      type="application/json"
    >
      []
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS___ext"
      type="application/json"
    >
      {"namedChunks":[]}
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[0]"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[1]"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[2]"
    >
    </script>
  </body>
</html>,
  "main-464f06daf5715e3d7e00.css": .external {
  display: none;
  font-size: 9px;
}
.qiwkl40 {
  border: 10px solid red;
}
.qiwkl41 {
  background: blue;
  font-size: 64px;
  height: 100px;
  width: 100px;
}
.ss2a6DC {
  display: flex;
}
.ss2a6DC .x5z4AMj {
  font-size: 32px;
}
,
  "main.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "runtime.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`styling start should start a development server 1`] = `
SCRIPTS: [
  "/runtime.js",
  "/2.js",
  "/main.js",
]
CSS: [
  "/main.css",
]
SOURCE HTML: <!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      My Awesome Project
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      data-chunk="main"
      rel="stylesheet"
      href="styles[0]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[0]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[1]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="scripts[2]"
    >
  </head>
  <body>
    <div id="app">
      <div>
        <div>
          Render type:
          Initial
        </div>
        <div class="styles__root___ss2a6DC">
          <span
            class="external"
            data-automation-external="true"
          >
            This should be invisible
          </span>
          <div
            class="styles__nested___x5z4AMj"
            data-automation-less="true"
          >
          </div>
          <div
            class="BlueBlock_root__qiwkl41 BlueBlock_border__qiwkl40"
            data-automation-vanilla="true"
          >
          </div>
        </div>
      </div>
    </div>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS__"
      type="application/json"
    >
      []
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS___ext"
      type="application/json"
    >
      {"namedChunks":[]}
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[0]"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[1]"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="scripts[2]"
    >
    </script>
  </body>
</html>
POST HYDRATE DIFFS: 
       <div>
         <div>
           Render type:
-          Initial
+          Client
         </div>
         <div class="styles__root___ss2a6DC">
           <span

`;
